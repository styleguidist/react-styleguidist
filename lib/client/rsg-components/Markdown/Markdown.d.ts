import React from 'react';
export declare const baseOverrides: {
    a: {
        component: React.SFC<{}>;
    };
    h1: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h2: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h3: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h4: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h5: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h6: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    p: {
        component: React.SFC<{}>;
        props: {
            semantic: string;
        };
    };
    em: {
        component: React.SFC<{}>;
        props: {
            semantic: string;
        };
    };
    strong: {
        component: React.SFC<{}>;
        props: {
            semantic: string;
        };
    };
    ul: {
        component: React.SFC<{}>;
    };
    ol: {
        component: React.SFC<{}>;
        props: {
            ordered: boolean;
        };
    };
    blockquote: {
        component: React.SFC<{}>;
    };
    code: {
        component: React.SFC<{}>;
    };
    pre: {
        component: React.SFC<{}>;
    };
    input: {
        component: React.SFC<{}>;
    };
    hr: {
        component: React.SFC<{}>;
    };
    table: {
        component: React.SFC<{}>;
    };
    thead: {
        component: React.SFC<{}>;
    };
    th: {
        component: React.SFC<{}>;
        props: {
            header: boolean;
        };
    };
    tbody: {
        component: React.SFC<{}>;
    };
    tr: {
        component: React.SFC<{}>;
    };
    td: {
        component: React.SFC<{}>;
    };
    details: {
        component: React.SFC<{}>;
    };
    summary: {
        component: React.SFC<{}>;
    };
};
export declare const inlineOverrides: {
    p: {
        component: React.ComponentType<Pick<import("../Text/TextRenderer").TextProps, string | number>>;
    };
    a: {
        component: React.SFC<{}>;
    };
    h1: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h2: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h3: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h4: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h5: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    h6: {
        component: React.SFC<{}>;
        props: {
            level: number;
        };
    };
    em: {
        component: React.SFC<{}>;
        props: {
            semantic: string;
        };
    };
    strong: {
        component: React.SFC<{}>;
        props: {
            semantic: string;
        };
    };
    ul: {
        component: React.SFC<{}>;
    };
    ol: {
        component: React.SFC<{}>;
        props: {
            ordered: boolean;
        };
    };
    blockquote: {
        component: React.SFC<{}>;
    };
    code: {
        component: React.SFC<{}>;
    };
    pre: {
        component: React.SFC<{}>;
    };
    input: {
        component: React.SFC<{}>;
    };
    hr: {
        component: React.SFC<{}>;
    };
    table: {
        component: React.SFC<{}>;
    };
    thead: {
        component: React.SFC<{}>;
    };
    th: {
        component: React.SFC<{}>;
        props: {
            header: boolean;
        };
    };
    tbody: {
        component: React.SFC<{}>;
    };
    tr: {
        component: React.SFC<{}>;
    };
    td: {
        component: React.SFC<{}>;
    };
    details: {
        component: React.SFC<{}>;
    };
    summary: {
        component: React.SFC<{}>;
    };
};
interface MarkdownProps {
    text: string;
    inline?: boolean;
}
export declare const Markdown: React.FunctionComponent<MarkdownProps>;
export default Markdown;
