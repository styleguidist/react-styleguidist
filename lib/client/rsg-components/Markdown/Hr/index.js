export { default } from 'rsg-components/Markdown/Hr/HrRenderer';