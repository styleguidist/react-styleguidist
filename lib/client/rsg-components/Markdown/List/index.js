export { default } from 'rsg-components/Markdown/List/ListRenderer';