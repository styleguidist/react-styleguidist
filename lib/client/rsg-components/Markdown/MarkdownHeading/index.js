export { default } from 'rsg-components/Markdown/MarkdownHeading/MarkdownHeadingRenderer';