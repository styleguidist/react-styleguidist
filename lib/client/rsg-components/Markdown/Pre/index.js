export { default } from 'rsg-components/Markdown/Pre/PreRenderer';
export * from 'rsg-components/Markdown/Pre/PreRenderer';