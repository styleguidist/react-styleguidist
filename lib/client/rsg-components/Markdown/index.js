export { default } from 'rsg-components/Markdown/Markdown';