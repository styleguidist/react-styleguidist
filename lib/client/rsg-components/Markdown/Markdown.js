import "core-js/modules/es.object.assign";
import React, { isValidElement } from 'react';
import PropTypes from 'prop-types';
import { compiler } from 'markdown-to-jsx';
import stripHtmlComments from 'strip-html-comments';
import Link from 'rsg-components/Link';
import Text from 'rsg-components/Text';
import Para from 'rsg-components/Para';
import MarkdownHeading from 'rsg-components/Markdown/MarkdownHeading';
import List from 'rsg-components/Markdown/List';
import Blockquote from 'rsg-components/Markdown/Blockquote';
import PreBase from 'rsg-components/Markdown/Pre';
import Code from 'rsg-components/Code';
import Checkbox from 'rsg-components/Markdown/Checkbox';
import Hr from 'rsg-components/Markdown/Hr';
import { Details, DetailsSummary } from 'rsg-components/Markdown/Details';
import { Table, TableHead, TableBody, TableRow, TableCell } from 'rsg-components/Markdown/Table';

var Pre = function Pre(props) {
  if (isValidElement(props.children)) {
    // Avoid rendering <Code> inside <Pre>
    return /*#__PURE__*/React.createElement(PreBase, props.children.props);
  }

  return /*#__PURE__*/React.createElement(PreBase, props);
};

Pre.propTypes = {
  children: PropTypes.node
};
export var baseOverrides = {
  a: {
    component: Link
  },
  h1: {
    component: MarkdownHeading,
    props: {
      level: 1
    }
  },
  h2: {
    component: MarkdownHeading,
    props: {
      level: 2
    }
  },
  h3: {
    component: MarkdownHeading,
    props: {
      level: 3
    }
  },
  h4: {
    component: MarkdownHeading,
    props: {
      level: 4
    }
  },
  h5: {
    component: MarkdownHeading,
    props: {
      level: 5
    }
  },
  h6: {
    component: MarkdownHeading,
    props: {
      level: 6
    }
  },
  p: {
    component: Para,
    props: {
      semantic: 'p'
    }
  },
  em: {
    component: Text,
    props: {
      semantic: 'em'
    }
  },
  strong: {
    component: Text,
    props: {
      semantic: 'strong'
    }
  },
  ul: {
    component: List
  },
  ol: {
    component: List,
    props: {
      ordered: true
    }
  },
  blockquote: {
    component: Blockquote
  },
  code: {
    component: Code
  },
  pre: {
    component: Pre
  },
  input: {
    component: Checkbox
  },
  hr: {
    component: Hr
  },
  table: {
    component: Table
  },
  thead: {
    component: TableHead
  },
  th: {
    component: TableCell,
    props: {
      header: true
    }
  },
  tbody: {
    component: TableBody
  },
  tr: {
    component: TableRow
  },
  td: {
    component: TableCell
  },
  details: {
    component: Details
  },
  summary: {
    component: DetailsSummary
  }
};
export var inlineOverrides = Object.assign({}, baseOverrides, {
  p: {
    component: Text
  }
});
export var Markdown = function Markdown(_ref) {
  var text = _ref.text,
      inline = _ref.inline;
  var overrides = inline ? inlineOverrides : baseOverrides;
  return compiler(stripHtmlComments(text), {
    overrides: overrides,
    forceBlock: true
  });
};
Markdown.propTypes = {
  text: PropTypes.string.isRequired,
  inline: PropTypes.bool
};
export default Markdown;