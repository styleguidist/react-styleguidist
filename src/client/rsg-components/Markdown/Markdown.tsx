import React, { isValidElement, PropsWithChildren } from 'react';
import PropTypes from 'prop-types';
import { compiler } from 'markdown-to-jsx';
import stripHtmlComments from 'strip-html-comments';
import Link from 'rsg-components/Link';
import Text from 'rsg-components/Text';
import Para from 'rsg-components/Para';
import MarkdownHeading from 'rsg-components/Markdown/MarkdownHeading';
import List from 'rsg-components/Markdown/List';
import Blockquote from 'rsg-components/Markdown/Blockquote';
import PreBase, { PreProps } from 'rsg-components/Markdown/Pre';
import Code from 'rsg-components/Code';
import Checkbox from 'rsg-components/Markdown/Checkbox';
import Hr from 'rsg-components/Markdown/Hr';
import { Details, DetailsSummary } from 'rsg-components/Markdown/Details';
import { Table, TableHead, TableBody, TableRow, TableCell } from 'rsg-components/Markdown/Table';

const Pre = (props: PreProps) => {
	if (isValidElement(props.children)) {
		// Avoid rendering <Code> inside <Pre>
		return <PreBase {...props.children.props} />;
	}
	return <PreBase {...props} />;
};
Pre.propTypes = {
	children: PropTypes.node,
};

export const baseOverrides = {
	a: {
		component: Link as React.FC,
	},
	h1: {
		component: MarkdownHeading as React.FC,
		props: {
			level: 1,
		},
	},
	h2: {
		component: MarkdownHeading as React.FC,
		props: {
			level: 2,
		},
	},
	h3: {
		component: MarkdownHeading as React.FC,
		props: {
			level: 3,
		},
	},
	h4: {
		component: MarkdownHeading as React.FC,
		props: {
			level: 4,
		},
	},
	h5: {
		component: MarkdownHeading as React.FC,
		props: {
			level: 5,
		},
	},
	h6: {
		component: MarkdownHeading as React.FC,
		props: {
			level: 6,
		},
	},
	p: {
		component: Para as React.FC,
		props: {
			semantic: 'p',
		},
	},
	em: {
		component: Text as React.FC,
		props: {
			semantic: 'em',
		},
	},
	strong: {
		component: Text as React.FC,
		props: {
			semantic: 'strong',
		},
	},
	ul: {
		component: List as React.FC,
	},
	ol: {
		component: List as React.FC,
		props: {
			ordered: true,
		},
	},
	blockquote: {
		component: Blockquote as React.FC,
	},
	code: {
		component: Code as React.FC,
	},
	pre: {
		component: Pre as React.FC<PropsWithChildren>,
	},
	input: {
		component: Checkbox as React.FC,
	},
	hr: {
		component: Hr as React.FC,
	},
	table: {
		component: Table as React.FC,
	},
	thead: {
		component: TableHead as React.FC,
	},
	th: {
		component: TableCell as React.FC,
		props: {
			header: true,
		},
	},
	tbody: {
		component: TableBody as React.FC,
	},
	tr: {
		component: TableRow as React.FC,
	},
	td: {
		component: TableCell as React.FC,
	},
	details: {
		component: Details as React.FC,
	},
	summary: {
		component: DetailsSummary as React.FC,
	},
};

export const inlineOverrides = {
	...baseOverrides,
	p: {
		component: Text,
	},
};

interface MarkdownProps {
	text: string;
	inline?: boolean;
}

export const Markdown: React.FunctionComponent<MarkdownProps> = ({ text, inline }) => {
	const overrides = inline ? inlineOverrides : baseOverrides;
	return compiler(stripHtmlComments(text), { overrides, forceBlock: true });
};

Markdown.propTypes = {
	text: PropTypes.string.isRequired,
	inline: PropTypes.bool,
};

export default Markdown;
