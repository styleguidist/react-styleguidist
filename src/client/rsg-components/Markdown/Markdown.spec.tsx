import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import Markdown from './Markdown';

describe('Markdown', () => {
	const expectSnapshotToMatch = (markdown: string) => {
		const actual = renderer.create(<Markdown text={markdown} />);

		expect(actual.toJSON()).toMatchSnapshot();
	};

	it('should forward DOM attributes onto resulting HTML', () => {
		const markdown =
			'<a href="test.com" id="preserve-my-id" class="preserve-my-class">Something</a>';

		const { getByRole } = render(<Markdown text={markdown} />);

		expect(getByRole('link').getAttribute('id')).toEqual('preserve-my-id');
		expect(getByRole('link').className).toContain('preserve-my-class');
	});

	it('should render links', () => {
		expectSnapshotToMatch('a [link](http://test.com)');
	});

	it('should render headings with generated ids', () => {
		expectSnapshotToMatch(`
# one
## two
### three
#### four
##### five
###### six
`);
	});

	it('should render paragraphs', () => {
		expectSnapshotToMatch(`
a paragraph

another paragraph
		`);
	});

	it('should render emphasis and strong text', () => {
		expectSnapshotToMatch(`
this text is **strong**

and this is _emphasized_
		`);
	});

	it('should render unordered lists', () => {
		expectSnapshotToMatch(`
* list
* item
* three
`);
	});

	it('should render ordered lists', () => {
		expectSnapshotToMatch(`
1. list
1. item
1. three
`);
	});

	it('should render mixed nested lists', () => {
		expectSnapshotToMatch(`
* list 1
* list 2
  1. Sub-list
  1. Sub-list
  1. Sub-list
* list 3
`);
	});

	it('should render check-lists', () => {
		expectSnapshotToMatch(`
* [ ] to do 1
* [ ] to do 2
* [x] to do 3
`);
	});

	it('should render a blockquote', () => {
		expectSnapshotToMatch(`
> This is a blockquote.
> And this is a second line.
`);
	});

	it('should render pre-formatted text', () => {
		expectSnapshotToMatch(`
    this is preformatted
    so is this
`);
	});

	it('should render code blocks without escaping', () => {
		expectSnapshotToMatch(`
\`\`\`html
<foo></foo>
\`\`\`
`);
	});

	it('should render inline code with escaping', () => {
		expectSnapshotToMatch('Foo `<bar>` baz');
	});

	it('should render a horizontal rule', () => {
		expectSnapshotToMatch(`---`);
	});

	it('should render a table', () => {
		expectSnapshotToMatch(`
| heading 1 | heading 2 |
| --------- | --------- |
| foo		| bar		|
| more foo	| more bar	|
`);
	});

	it.only('should ignore single line comments', () => {
		const markdown = `Hello World
<!-- This is a single line comment -->
`;
		const { queryByText } = render(<Markdown text={markdown} />);

		expect(queryByText('This is a single line comment')).toBe(null);
	});

	it('should ignore multiline comments', () => {
		const markdown = `Hello World
<!--
This is a
multiline
comment
-->
`;
		const { queryByText } = render(<Markdown text={markdown} />);

		expect(
			queryByText(`This is a
multiline
comment`)
		).toBe(null);
	});
});

describe('Markdown inline', () => {
	const expectSnapshotToMatch = (markdown: string) => {
		const actual = renderer.create(<Markdown text={markdown} inline />);

		expect(actual).toMatchSnapshot();
	};

	it('should render text in a span', () => {
		expectSnapshotToMatch('Hello world!');
	});
});
