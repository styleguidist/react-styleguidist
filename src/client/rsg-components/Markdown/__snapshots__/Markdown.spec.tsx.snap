// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown inline should render text in a span 1`] = `
<span className="rsg--text-11 rsg--inheritSize-12 rsg--baseColor-16">
  Hello world!
</span>
`;

exports[`Markdown should render a blockquote 1`] = `
<blockquote className="rsg--blockquote-25">
  <p className="rsg--para-2">
    This is a blockquote.
And this is a second line.
  </p>
</blockquote>
`;

exports[`Markdown should render a horizontal rule 1`] = `<hr className="rsg--hr-28">`;

exports[`Markdown should render a table 1`] = `
<table className="rsg--table-29">
  <thead className="rsg--thead-30">
    <tr>
      <th className="rsg--th-32 rsg--td-31">
        heading 1
      </th>
      <th className="rsg--th-32 rsg--td-31">
        heading 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td className="rsg--td-31">
        foo
      </td>
      <td className="rsg--td-31">
        bar
      </td>
    </tr>
    <tr>
      <td className="rsg--td-31">
        more foo
      </td>
      <td className="rsg--td-31">
        more bar
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Markdown should render check-lists 1`] = `
<ul className="rsg--list-21">
  <li className="rsg--li-23">
    <input type="checkbox"
           readonly
           className="rsg--input-24"
    >
    to do 1
  </li>
  <li className="rsg--li-23">
    <input type="checkbox"
           readonly
           className="rsg--input-24"
    >
    to do 2
  </li>
  <li className="rsg--li-23">
    <input type="checkbox"
           checked
           readonly
           className="rsg--input-24"
    >
    to do 3
  </li>
</ul>
`;

exports[`Markdown should render code blocks without escaping 1`] = `
<pre className="lang-html rsg--pre-26">
  <foo>
  </foo>
</pre>
`;

exports[`Markdown should render emphasis and strong text 1`] = `
<div>
  <p className="rsg--para-2">
    this text is
    <strong className="rsg--text-11 rsg--inheritSize-12 rsg--baseColor-16 rsg--strong-19">
      strong
    </strong>
  </p>
  <p className="rsg--para-2">
    and this is
    <em className="rsg--text-11 rsg--inheritSize-12 rsg--baseColor-16 rsg--em-18">
      emphasized
    </em>
  </p>
</div>
`;

exports[`Markdown should render headings with generated ids 1`] = `
<div>
  <div className="rsg--spacing-3">
    <h1 id="one"
        className="rsg--heading-4 rsg--heading1-5"
    >
      one
    </h1>
  </div>
  <div className="rsg--spacing-3">
    <h2 id="two"
        className="rsg--heading-4 rsg--heading2-6"
    >
      two
    </h2>
  </div>
  <div className="rsg--spacing-3">
    <h3 id="three"
        className="rsg--heading-4 rsg--heading3-7"
    >
      three
    </h3>
  </div>
  <div className="rsg--spacing-3">
    <h4 id="four"
        className="rsg--heading-4 rsg--heading4-8"
    >
      four
    </h4>
  </div>
  <div className="rsg--spacing-3">
    <h5 id="five"
        className="rsg--heading-4 rsg--heading5-9"
    >
      five
    </h5>
  </div>
  <div className="rsg--spacing-3">
    <h6 id="six"
        className="rsg--heading-4 rsg--heading6-10"
    >
      six
    </h6>
  </div>
</div>
`;

exports[`Markdown should render inline code with escaping 1`] = `
<p className="rsg--para-2">
  Foo
  <code className="rsg--code-27">
    &lt;bar&gt;
  </code>
  baz
</p>
`;

exports[`Markdown should render links 1`] = `
<p className="rsg--para-2">
  a
  <a href="http://test.com"
     className="rsg--link-0"
  >
    link
  </a>
</p>
`;

exports[`Markdown should render mixed nested lists 1`] = `
<ul className="rsg--list-21">
  <li className="rsg--li-23">
    list 1
  </li>
  <li className="rsg--li-23">
    list 2
    <ol className="rsg--list-21 rsg--ordered-22">
      <li className="rsg--li-23">
        Sub-list
      </li>
      <li className="rsg--li-23">
        Sub-list
      </li>
      <li className="rsg--li-23">
        Sub-list
      </li>
    </ol>
  </li>
  <li className="rsg--li-23">
    list 3
  </li>
</ul>
`;

exports[`Markdown should render ordered lists 1`] = `
<ol className="rsg--list-21 rsg--ordered-22">
  <li className="rsg--li-23">
    list
  </li>
  <li className="rsg--li-23">
    item
  </li>
  <li className="rsg--li-23">
    three
  </li>
</ol>
`;

exports[`Markdown should render paragraphs 1`] = `
<div>
  <p className="rsg--para-2">
    a paragraph
  </p>
  <p className="rsg--para-2">
    another paragraph
  </p>
</div>
`;

exports[`Markdown should render pre-formatted text 1`] = `
<pre className="rsg--pre-26">
  this is preformatted
so is this
</pre>
`;

exports[`Markdown should render unordered lists 1`] = `
<ul className="rsg--list-21">
  <li className="rsg--li-23">
    list
  </li>
  <li className="rsg--li-23">
    item
  </li>
  <li className="rsg--li-23">
    three
  </li>
</ul>
`;
